resources/aws_ec2.go

```diff
@@ -158,60 +200,36 @@ func (obj *AwsEc2Res) Init() error {
 func (obj *AwsEc2Res) Watch() error {
 	send := false
 	var exit *error
-	if err := obj.Running(); err != nil {
-		return err
-	}
-	type chanStruct struct {
-		str string
-		err error
-	}
-	awsChan := make(chan *chanStruct)
-	closeChan := make(chan struct{})
+	obj.awsChan = make(chan *chanStruct)
+	obj.closeChan = make(chan struct{})
 	ctx, cancel := context.WithCancel(context.TODO())
 	wg := &sync.WaitGroup{}
 	wg.Add(1)
 	go func() {
 		defer wg.Done()
 		select {
-		case <-closeChan:
+		case <-obj.closeChan:
 			cancel()
 		}
 	}()
-	wg.Add(1)
-	defer wg.Wait()
-	go func() {
-		defer wg.Done()
-		defer close(awsChan)
-		for {
-			diInput := &ec2.DescribeInstancesInput{
-				Filters: []*ec2.Filter{
-					{
-						Name:   aws.String("tag:Name"),
-						Values: []*string{aws.String(obj.prependName())},
-					},
-					{
-						Name: aws.String("instance-state-name"),
-						Values: []*string{
-							aws.String("pending"),
-							aws.String("running"),
-							aws.String("stopping"),
-							aws.String("stopped"),
-						},
-					},
-				},
-			}
-			diOutput, err := obj.client.DescribeInstances(diInput)
-			if err != nil {
-				select {
-				case awsChan <- &chanStruct{
-					err: errwrap.Wrapf(err, "error describing instances"),
-				}:
-				case <-closeChan:
-				}
-				return
-			}
-			if obj.State == "running" {
-				stoppedInput := &ec2.DescribeInstancesInput{
+	if obj.SnsEndpoint != "" {
```

Comment:
below here in `Watch()` patch is basically unreadable as a single commit. Might have to review individual commits.
If I understand correctly: We're basically doing Watch type (1) or Watch type (2) and they otherwise don't share much code, correct? Where (1) is the long poll, (2) is the sns thing, and maybe in the future (3) is an sqs long poller. If that's the case, it's probably best to have three top level Watch methods, like `WatchLongPoll`, `WatchSns` and `WatchSqs` and the main watch calls the appropriate one as needed.